[13.0][ADD] hr_leave_attach_autovalidated // [FIX] hr_leave_has_attachment: allow everyone to edit attachments

IriaAlonso opened the pull request:
Allow to attach files on autovalidated absences and allow everyone to edit attachments

IriaAlonso commented:
update commit description, @ChristianSantamaria can you review it?

ChristianSantamaria commented:
Not working in my local enviroment
![imagen](https://github.com/user-attachments/assets/c70d6a93-9a0b-461c-a03d-fc8659545c25)

IriaAlonso commented:
@ChristianSantamaria can you review it?

ChristianSantamaria commented:
@dalonsod LGTM tested in local and alu enviroment :+1:

dalonsod commented on hr_leave_attach_autovalidated_absence/models/ir_attachment.py:
```diff
@@ -0,0 +1,40 @@
+                    hr_leave_id.validation_type == 'no_validation' and
+                    hr_leave_id.state == 'validate'
+                ):
+                    return super(IrAttachment, self.sudo()).create(vals_list)
+                else:
+                    return super(IrAttachment, self).create(vals_list)
+            else:
+                return super(IrAttachment, self).create(vals_list)
```
This code should work for a single record, but not for a truly `multi` call (`for` loop is actually called once)

dalonsod commented on hr_leave_attach_autovalidated_absence/models/ir_attachment.py:
```diff
@@ -0,0 +1,40 @@
+                        hr_leave_id.validation_type == 'no_validation' and
+                        hr_leave_id.state == 'validate'
+                    ):
+                        return super(IrAttachment, self.sudo()).unlink()
+                else:
+                    return super(IrAttachment, self).unlink()
+            else:
+                return super(IrAttachment, self).unlink()
```
the same as above.

dalonsod reviewed (changes requested):
See comments

IriaAlonso committed:
[ADD] hr_leave_attach_autovalidated
Allow to attach files on autovalidated absences.

IriaAlonso committed:
[FIX] hr_leave_has_attachment: allow everyone to edit attachments
The 'hr_leave_attach_autovalidated' addon allow to delete an attachment,
but this addon modify the unlink() method editing the attachment, giving permissions error.
Adding the sudo() allow to edit the attachment, fixing this problem.

IriaAlonso commented:
@dalonsod changes done

dalonsod commented on hr_leave_attach_autovalidated_absence/models/ir_attachment.py:
```diff
@@ -0,0 +1,42 @@
+                        hr_leave_id.employee_id.user_id.id == self.env.user.id and
+                        hr_leave_id.validation_type == 'no_validation' and
+                        hr_leave_id.state == 'validate'
+                    ):
+                        res = super(IrAttachment, self.sudo()).unlink()
+            else:
+                res = super(IrAttachment, self).unlink()
```
`self` is the whole object. As you're iterating, you should make each individual call to `unlink` using every iteration object:
```suggestion
                    ):
                        res = super(IrAttachment, attach.sudo()).unlink()
            else:
                res = super(IrAttachment, attach).unlink()
```

IriaAlonso commented:
@dalonsod done

dalonsod commented:
OK, @IriaAlonso could you rename this? E.g. to `hr_leave_attach_autovalidated`, @ChristianSantamaria what do you think?

ChristianSantamaria commented:
@dalonsod I like it :+1:

IriaAlonso commented:
@dalonsod name change

dalonsod commented:
@IriaAlonso take a look at file tree, there are some files that shouldn't belong to this PR

ChristianSantamaria commented:
:+1: Tested in alu enviroment

dalonsod commented:
@IriaAlonso please edit first comment of this PR, as final PR content is different than original one and now includes two improvements/fixes
